src/types/weakref/reference.rs

```diff
@@ -0,0 +1,1200 @@
+use crate::err::PyResult;
+use crate::ffi_ptr_ext::FfiPtrExt;
+use crate::type_object::{PyTypeCheck, PyTypeInfo};
+use crate::types::any::PyAnyMethods;
+use crate::{ffi, Borrowed, Bound, PyAny, PyNativeType, Python, ToPyObject};
+
+/// Represents a Python `weakref.ReferenceType`.
+///
+/// In Python this is created by calling `weakref.ref`.
+#[repr(transparent)]
+pub struct PyWeakRef(PyAny);
+
+pyobject_native_type!(
+    PyWeakRef,
+    ffi::PyWeakReference,
+    pyobject_native_static_type_object!(ffi::_PyWeakref_RefType),
+    #module=Some("weakref"),
+    #checkfunction=ffi::PyWeakref_CheckRefExact
+);
+
+impl PyWeakRef {
+    /// Deprecated form of [`PyWeakRef::new_bound`].
+    #[inline]
+    #[track_caller]
+    #[cfg_attr(
+        not(feature = "gil-refs"),
+        deprecated(
+            since = "0.21.0",
+            note = "`PyWeakRef::new` will be replaced by `PyWeakRef::new_bound` in a future PyO3 version"
+        )
+    )]
+    pub fn new<T>(py: Python<'_>, object: T) -> PyResult<&'_ PyWeakRef>
+    where
+        T: ToPyObject,
+    {
+        Self::new_bound(py, object).map(Bound::into_gil_ref)
+    }
+
+    /// Constructs a new Weak Reference (`weakref.ref`/`weakref.ReferenceType`) for the given object.
+    ///
+    /// Returns a `TypeError` if `object` is not weak referenceable (Most native types and PyClasses without `weakref` flag).
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use pyo3::prelude::*;
+    /// use pyo3::types::PyWeakRef;
+    ///
+    /// #[pyclass(weakref)]
+    /// struct Foo { /* fields omitted */ }
+    ///
+    /// # fn main() -> PyResult<()> {
+    /// Python::with_gil(|py| {
+    ///     let foo = Bound::new(py, Foo {})?;
+    ///     let weakref = PyWeakRef::new_bound(py, foo.clone())?;
+    ///     assert!(
+    ///         // In normal situations where a direct `Bound<'py, Foo>` is required use `upgrade::<Foo>`
+    ///         weakref.get_object()
+    ///             .map_or(false, |obj| obj.is(&foo))
+    ///     );
+    ///
+    ///     let weakref2 = PyWeakRef::new_bound(py, foo.clone())?;
+    ///     assert!(weakref.is(&weakref2));
+    ///
+    ///     drop(foo);
+    ///
+    ///     assert!(weakref.get_object().is_none());
+    ///     Ok(())
+    /// })
+    /// # }
+    /// ```
+    #[track_caller]
+    pub fn new_bound<T>(py: Python<'_>, object: T) -> PyResult<Bound<'_, PyWeakRef>>
+    where
+        T: ToPyObject,
+    {
+        unsafe {
+            Bound::from_owned_ptr_or_err(
+                py,
+                ffi::PyWeakref_NewRef(object.to_object(py).as_ptr(), ffi::Py_None()),
```

Comment:
`object.to_object(py).as_ptr()` is dangling!

At the very least we need do `let object = object.to_object()`. You might want to check out the `fn inner` patterns we have in e.g. `PyAnyMethods`, they help avoid generic code bloat.

A related question, would be do we need to allow `T: ToPyObject` for this function, or should we just take `&Bound<PyAny>`? It seems strange to me to create a new Python object and then immediately pass it to `weakref.ref`, presumably the GC would immediately delete it.

Reply:
It would probably indeed make more sense to take a `&Bound<PyAny>`, but wouldn't a `&Bound<T>` result in more flexibility?

Reply:
I've done something like the inner pattern, for the `callback` I didn't know how what would be wise to do (taking `Bound` or not). So I've left that for now.

Reply:
It currently takes `&Bound<PyAny>` for all arguments except for `callback` arguments, since the garbage collector should will (most likely) keep the callback if the  `weakref` exists.

However, I'm also open to changing it to `&Bound<PyAny>` if it would make more sense.